src/code/sched.c

```diff
@@ -165,48 +230,75 @@ OSScTask* func_800C89D4(SchedContext* sc, OSScTask* task) {
         return NULL;
     }
 
-    if ((sc->pendingSwapBuf2 != NULL ? sc->pendingSwapBuf2->swapBuffer : NULL) == task->framebuffer->fb1) {
+    // If the task's framebuffer is one of the pending swaps or NULL.
+    // In conjunction with the above, these checks are redundant as the pending swap buffers will only be
+    // NULL here, so these could have been simplified to checks for the task's framebuffer being non-NULL.
+
+    if (((sc->pendingSwapBuf2 != NULL) ? sc->pendingSwapBuf2->swapBuffer : NULL) == task->framebuffer->framebuffer) {
         return NULL;
     }
 
-    if ((sc->pendingSwapBuf1 != NULL ? sc->pendingSwapBuf1->swapBuffer : NULL) == task->framebuffer->fb1) {
+    if (((sc->pendingSwapBuf1 != NULL) ? sc->pendingSwapBuf1->swapBuffer : NULL) == task->framebuffer->framebuffer) {
         return NULL;
     }
 
-    if (osViGetCurrentFramebuffer() == (u32*)task->framebuffer->fb1) {
+    // If the task's framebuffer is the current framebuffer, abort
+
+    if (osViGetCurrentFramebuffer() == task->framebuffer->framebuffer) {
         return NULL;
     }
 
     return task;
 }
 
-s32 Sched_Schedule(SchedContext* sc, OSScTask** sp, OSScTask** dp, s32 state) {
+/**
+ * Schedules the next tasks to run on the RSP and RDP
+ *
+ * @param sc Scheduler
+ * @param spTaskOut Next task to run on the RSP
+ * @param dpTaskOut Next task to run on the RDP
+ * @param state Bits containing whether the RSP and RDP are currently in use
+ * @return Bits containing whether the RSP and RDP will be in use after starting the next tasks
+ */
+s32 Sched_Schedule(SchedContext* sc, OSScTask** spTaskOut, OSScTask** dpTaskOut, s32 state) {
     s32 ret = state;
     OSScTask* gfxTask = sc->gfxListHead;
     OSScTask* audioTask = sc->audioListHead;
 
-    if (sc->doAudio && (ret & OS_SC_SP)) {
-        *sp = audioTask;
+    if (sc->doAudio && (state & OS_SC_SP)) {
```

Comment:
Yeah, you're right that this isn't a great way to do things and should be highlighted in a comment. The system works since there's a long enough time between sending audio tasks that the earlier task will have been sent to the RSP by then.

As for possible fixes, `sc->audioListHead != NULL` is just as good for the highlighted case, however `doAudio` is also used to check whether the scheduler should request the current RSP task to yield. The yield request could probably be sent in `Sched_QueueTask` without any issues.

Reply:
Why would checking a non empty audio task list not be good enough to ask the current rsp task to yield?